artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -1130,9 +1136,14 @@ public void addHead(final MessageReference ref, boolean scheduling) {
    /* Called when a message is cancelled back into the queue */
    @Override
    public void addSorted(final MessageReference ref, boolean scheduling) {
+      if (RefCountMessage.isTraceEnabled()) {
+         RefCountMessage.deferredDebug(ref.getMessage(), "returning to queue {}", this.getAddress());
```

Comment:
There are a few "returning to queue" events. Although they can be separated by the stacktrace detail it would probably be nicer to make the text more distinct to begin with, and help out the poor sod digging through streams of stacktraces later.